### DISCUSSION TITLE
[Proposal]: New operator for rotate shift

### DISCUSSION THREAD

**Discussion by anchurcn** [dotnet/csharplang#4685] [Language Ideas]
# New operator for rotate shift

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary

Bits operation is common in primitive integers.
There have been SHR(Shift Right) operator and SHL(Shift Left) operator for unsigned integers.
```csharp
uint value = 0xAABBCCDD;
uint output = value >> 8;
// output is 0x00AABBCC
```
Also  SAR(Shift Arithmetic Right) operator and SAL(Shift Arithmetic Left) operator are for signed integers.
```csharp
int value = (int)0xFFAABBCC;
int output = value >> 8;
// output is 0xFFFFAABB
```
But there are no operators for rotate shift operations.

## Motivation
[motivation]: #motivation
If we want transform 0xAABBCCDD into 0xDDAABBCC,
we need to write a long expression.
```csharp
uint value = 0xAABBCCDD;
int nBits = sizeof(value) * 8;// value's width in bits
int n=8;// shift number of bits
uint output = (value << (nBits - n) | value >> n)
//output is 0xDDAABBCC
```
If new operators come, we have a more elegant way.
Assume >>> is a new operator for ROR(Rotate Right) operation.
```csharp
uint value = 0xAABBCCDD;
uint output = value >>> 8;
// output is 0xDDAABBCC
```

## Detailed design
[design]: #detailed-design

IL may not have direct instructions for ROR or ROL, we can just express it as 
(value << (nBits - n) | value >> n) or (value << (nBits - n) | value >> n) in IL.
But I think JIT compiler will generate the best assembly for the target platform.

## Drawbacks
[drawbacks]: #drawbacks

## Alternatives
[alternatives]: #alternatives

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by CyrusNajmabadi** [comment]
This def feels like something a method could be exposed for.  It could presumably be implemented by the jit with whatever smarts are needed.

**Reply by alexrp** [reply]
They even exist already: https://docs.microsoft.com/en-us/dotnet/api/system.numerics.bitoperations (and are recognized by the JIT)

**Reply by anchurcn** [reply]
Yes, method can help a lot. I have made some extension method for my project. But operator is better I think.
